### ISSUE TITLE
Proposal: Unify the interfaces of the data structures in standard library 

### ISSUE THREAD

**Issue by glyh** [ziglang/zig#8233]
I've tried some std data structures like ArrayLIsts, TailQueue, AutoHashMap. 

However, each one seems to have it's own way of doing things. 

For example, when you iterate through these containers, you do it differently : 
In ArrayLists, you do this:

```zig
var i: u32 = 0; 
while (i < a.items.len) : (i += 1) { 
     //...
}
```

In TailQueue, you do this: 
```zig
var it = a.first; 
while (it) |node| : (it = node.next) {
    //...
}
```

In AutoHashMap, you do this: 
```zig
var it = a.iterator();
while (it.next()) |kv| {
    //...
} 
```

As zig says in its zen: "Only one obvious way to do things." and "Reduce the amount one must remember.", I think we may need some uniform convention for operating on containers.

**Comment by glyh** [comment]
One more thing to mention, the data structures' name in Zig are also confusing, we have SingleLinkedList refers to one way linked list, while TailQueue refers to bi-directional linked list. This actually brings confusing.

**Comment by glyh** [comment]
A good example is Clojure's collection abstraction. they have a series of methods working on an abstraction queue, as long as they have some base function implemented: first, next, etc. This identical design make it really easy for programmers to do design their programs based on std.

**Comment by glyh** [comment]
A more wired thing is that TailQueue don't actually received allocators, it expects you to use allocators and alloc some memory, init something and pass the address to the LinkList, it's pretty different from other data structures who receive allocators in their construction functions.

**Comment by tauoverpi** [comment]
TailQueue doesn't need an allocator as you're not always storing your node in the heap. An example would be  https://github.com/ziglang/zig/blob/ce14bc7176f9e441064ffdde2d85e35fd78977f2/lib/std/event/loop.zig#L769-L780 which stores the node in the frame of the suspending function instead of needing to use an allocator for it.

Though a unified iterator interface similar to how there's Reader and Writer would be nice.

**Comment by komuw** [comment]
> One more thing to mention, the data structures' name in Zig are also confusing, we have SingleLinkedList refers to one way linked list, while TailQueue refers to bi-directional linked list. This actually brings confusement.


https://freenode.irclog.whitequark.org/zig/2019-06-10 (`ctrl f` TailQueue on that page)  

also you may want to cc `daurnimator`

**Comment by locriacyber** [comment]
Some thoughts
1. Tiered support for standard library. e.g. File access API should be more stable than Data Structures
2. [Support for polymorphism in standard library](https://github.com/ziglang/zig/issues/10184)

**Comment by jayschwa** [comment]
Related issue: #1629

**Comment by jayschwa** [comment]
> One more thing to mention, the data structures' name in Zig are also confusing, we have SingleLinkedList refers to one way linked list, while TailQueue refers to bi-directional linked list. This actually brings confusement.

I agree, I just got mildly tripped up by this. I searched the stdlib autodocs for "linked" and was surprised to see only `SinglyLinkedList` come up. It wasn't until I opened the source file that I realized `TailQueue` is the doubly-linked list implementation. Autodocs could perhaps beef up search by looking in the doc comments, but it still seems inconsistent that `SinglyLinkedList` is named according to the implementation and `TailQueue` is named according to the abstraction.

**Comment by andrewrk** [comment]
I agree, I dislike the name TailQueue.

**Comment by andrewrk** [comment]
partially a duplicate of #1629, partially a rejected proposal
